pkg/cmd/roachtest/tests/mixed_version_sql_stats.go

```diff
@@ -0,0 +1,234 @@
+		// We'll validate that hitting the in-memory tables and either one of the
+		// persisted (full or cached activity) does not return error on any node.
+		// Ideally we'd have prior data in the persisted tables already so that we
+		// can easily query data from the previous aggregation interval. For now
+		// we'll just have to rely on waiting for the flush to occur.
+		expectedTableCount := 2
+
+		// Track the source tables used to service each request.
+		foundStmtTableSources := make(map[string]interface{})
+		foundTxnTableSources := make(map[string]interface{})
+
+		addToTableSources := func(res *serverpb.StatementsResponse) error {
+			if res.StmtsSourceTable != "" {
+				foundStmtTableSources[res.StmtsSourceTable] = struct{}{}
+			}
+			if res.TxnsSourceTable != "" {
+				foundTxnTableSources[res.TxnsSourceTable] = struct{}{}
+			}
+			return nil
+		}
+
+		// These are the two types of stats pages that power the admin UI.
+		requestTypes := []serverpb.CombinedStatementsStatsRequest_StatsType{
+			serverpb.CombinedStatementsStatsRequest_StmtStatsOnly,
+			serverpb.CombinedStatementsStatsRequest_TxnStatsOnly,
+		}
+
+		s := createSQLStatsRequestHelper(c, l)
+		// First, we'll attempt to get stats from the in-memory tables.
+		// We can ensure we request from in-memory tables by requesting stats from
+		// a time range for which no data is available, as the in-memory table
+		// is only used when no data is available from system tables.
+		for _, statsType := range requestTypes {
+			if err := s.requestSQLStatsFromEmptyInterval(ctx, roachNodes, statsType, addToTableSources); err != nil {
+				t.Fatal(err)
+			}
+		}
+		if foundStmtTableSources[server.CrdbInternalStmtStatsCombined] == nil || foundTxnTableSources[server.CrdbInternalTxnStatsCombined] == nil {
+			t.Fatal("expected to find in-memory tables in response, found %v, %v", foundStmtTableSources, foundTxnTableSources)
+		}
+
+		// Now we'll wait for the flush to occur and request stats from the persisted tables.
+		var timer timeutil.Timer
+		defer timer.Stop()
+		timer.Reset(5 * time.Second)
+
+		for len(foundStmtTableSources) < expectedTableCount || len(foundTxnTableSources) < expectedTableCount {
+			select {
+			case <-timer.C:
+				timer.Read = true
+				for _, statsType := range requestTypes {
+					// Requesting data from the last two hours should return data from the persisted tables eventually.
+					if err := s.requestSQLStatsFromLastTwoHours(ctx, roachNodes, statsType, addToTableSources); err != nil {
+						t.Fatal(err)
+					}
+				}
+			case <-ctx.Done():
+				return nil
+			}
+
+			l.Printf("waiting for flushed stats...")
+			timer.Reset(30 * time.Second)
+		}
+
+		return nil
+	}
+
+	mvt.OnStartup("set-cluster-settings", setClusterSettings)
+	stopTpcc := mvt.Workload("tpcc", workloadNode, initWorkload, runWorkload)
+	mvt.InMixedVersion("request-sql-stats", runGetRequests)
```

Comment:
It could be nice to split this step into a separate step for each type of request. Something like (high level):

```go
requestTypes = map[string]serverpb.CombinedStatementsStatsRequest_StatsType {
    "stmts stats": serverpb.CombinedStatementsStatsRequest_StmtStatsOnly,
    "txn stats": serverpb.CombinedStatementsStatsRequest_TxnStatsOnly,
}

for name, reqType := range requestTypes {
    mvt.InMixedVersion(fmt.Sprintf("request %s", name), runGetRequests(reqType))
}
```

The main advantage would be to explore different combinations of checking these types of requests in each test run, including checking them concurrently.

It would have the nice side effect of immediately knowing what kind of request failed. In the current implementation, since it's behind a loop, the stack trace alone doesn't allow you to tell what request type failed.